packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart

```diff
@@ -164,46 +166,55 @@ class ResidentWebRunner extends ResidentRunner {
-      _webFs = await webFsFactory(
-        target: target,
-        flutterProject: flutterProject,
-        buildInfo: debuggingOptions.buildInfo,
-        initializePlatform: debuggingOptions.initializePlatform,
-        hostname: debuggingOptions.hostname,
-        port: debuggingOptions.port,
-        skipDwds: device is WebServerDevice || !debuggingOptions.buildInfo.isDebug,
-      );
-      // When connecting to a browser, update the message with a seemsSlow notification
-      // to handle the case where we fail to connect.
-      if (debuggingOptions.browserLaunch) {
-        buildStatus.stop();
-        buildStatus = logger.startProgress(
-          'Attempting to connect to browser instance..',
-          timeout: const Duration(seconds: 30),
+      final int result = await asyncGuard(() async {
+        _webFs = await webFsFactory(
+          target: target,
+          flutterProject: flutterProject,
+          buildInfo: debuggingOptions.buildInfo,
+          initializePlatform: debuggingOptions.initializePlatform,
+          hostname: debuggingOptions.hostname,
+          port: debuggingOptions.port,
+          skipDwds: device is WebServerDevice || !debuggingOptions.buildInfo.isDebug,
         );
-      }
-      await device.startApp(package,
-        mainPath: target,
-        debuggingOptions: debuggingOptions,
-        platformArgs: <String, Object>{
-          'uri': _webFs.uri,
-        },
-      );
-      if (supportsServiceProtocol) {
-        _connectionResult = await _webFs.connect(debuggingOptions);
-        unawaited(_connectionResult.debugConnection.onDone.whenComplete(exit));
-      }
-    } catch (err) {
-      throwToolExit('Failed to build application for the web.');
+        // When connecting to a browser, update the message with a seemsSlow notification
+        // to handle the case where we fail to connect.
+        if (debuggingOptions.browserLaunch) {
+          buildStatus.stop();
+          buildStatus = logger.startProgress(
+            'Attempting to connect to browser instance..',
+            timeout: const Duration(seconds: 30),
+          );
+        }
+        await device.startApp(package,
+          mainPath: target,
+          debuggingOptions: debuggingOptions,
+          platformArgs: <String, Object>{
+            'uri': _webFs.uri,
+          },
+        );
+        if (supportsServiceProtocol) {
+          _connectionResult = await _webFs.connect(debuggingOptions);
+          unawaited(_connectionResult.debugConnection.onDone.whenComplete(() => exit(0)));
+        }
+        appStartedCompleter?.complete();
+        return attach(
+          connectionInfoCompleter: connectionInfoCompleter,
+          appStartedCompleter: appStartedCompleter,
+        );
+      });
+      return result;
+    } on WebSocketException {
+      throwToolExit('Failed to connect to WebSocket.');
+    } on Exception catch (err) {
```

Comment:
Throw/catch a more specific exception type here to avoid burying things we'd still like to bubble up to crash logging.